pandas/tseries/holiday.py

```diff
@@ -220,9 +220,15 @@ class from pandas.tseries.offsets
         self.day = day
         self.offset = offset
         self.start_date = (
-            Timestamp(start_date) if start_date is not None else start_date
+            Timestamp(start_date)
+            if start_date is not None
+            else AbstractHolidayCalendar.start_date
```

Comment:
I think this fix may be too specific.

After a quick look, I think the original example is returning an `Index` instead of a `DatetimeIndex` because the `concat` of empty data is a `Index` here

https://github.com/pandas-dev/pandas/blob/e7e5df5167be28dfc00c900d24bb13c9827524d0/pandas/tseries/holiday.py#L460

Reply:
This is just fractally weird, the more I look at it:

If you have both

- A list of rules in your calendar that includes At Least One holiday with an arbitrary start date defined (such as Juneteenth)
- A pair of dates at least 3 years before the earliest start date of that holiday that in turn includes no holidays

... then this behavior appears to recur for attempts to build the list of `holidays()` within a given range that does not include observed holidays on your `AbstractHolidayCalendar` of choice.

```
   ...: start=datetime.date(2018,8,1)
   ...: end=datetime.date(2018,8,31)

In [2]: cal1 = USFederalHolidayCalendar()

In [3]: cal1.holidays(start,end)
Out[3]: Index([], dtype='object')

In [4]: [ rule.dates(start,end) for rule in cal1.rules ]
Out[4]: 
[DatetimeIndex([], dtype='datetime64[ns]', freq=None),
 DatetimeIndex([], dtype='datetime64[ns]', freq=None),
 DatetimeIndex([], dtype='datetime64[ns]', freq=None),
 DatetimeIndex([], dtype='datetime64[ns]', freq=None),
 Index([], dtype='object'),
 DatetimeIndex([], dtype='datetime64[ns]', freq=None),
 DatetimeIndex([], dtype='datetime64[ns]', freq=None),
 DatetimeIndex([], dtype='datetime64[ns]', freq=None),
 DatetimeIndex([], dtype='datetime64[ns]', freq=None),
 DatetimeIndex([], dtype='datetime64[ns]', freq=None),
 DatetimeIndex([], dtype='datetime64[ns]', freq=None)]

In [5]: cal2 = USFederalHolidayCalendar()

In [6]: cal2.rules.pop(4)
Out[6]: Holiday: Juneteenth National Independence Day (month=6, day=19, observance=<function nearest_workday at 0x7f7d1e607910>)

In [7]: cal2.holidays(start,end)
Out[7]: DatetimeIndex([], dtype='datetime64[ns]', freq=None)

In [8]: [ rule.dates(start,end) for rule in cal2.rules ]
Out[8]: 
[DatetimeIndex([], dtype='datetime64[ns]', freq=None),
 DatetimeIndex([], dtype='datetime64[ns]', freq=None),
 DatetimeIndex([], dtype='datetime64[ns]', freq=None),
 DatetimeIndex([], dtype='datetime64[ns]', freq=None),
 DatetimeIndex([], dtype='datetime64[ns]', freq=None),
 DatetimeIndex([], dtype='datetime64[ns]', freq=None),
 DatetimeIndex([], dtype='datetime64[ns]', freq=None),
 DatetimeIndex([], dtype='datetime64[ns]', freq=None),
 DatetimeIndex([], dtype='datetime64[ns]', freq=None),
 DatetimeIndex([], dtype='datetime64[ns]', freq=None)]
```